### ISSUE TITLE
Detekt reports configuration is ignored when referenced from a buildscript

### ISSUE THREAD

**Issue by ahibrahimleague** [detekt/detekt#6141]
I'm applying a Gradle script at the root project that applies Detekt plugin to subprojects, and disables XML, Sarif but keeping the HTML reports.

## Expected Behavior
XML and Sarif reports shouldn't be generated.

## Observed Behavior
XML and Sarif reports are still generated.

## Steps to Reproduce
Below is a reproducer project that I've generated that showcase the issue.

Run `./gradlew detektMain`, and the build will fail with rules violations. Check the `app/build/reports/detekt` dir, and you'll notice XML and Sarif files being there while they shouldn't. Even though I disabled them.

The whole detekt.gradle script that's applied to the root project.

```
import io.gitlab.arturbosch.detekt.Detekt

buildscript {
    repositories {
        mavenCentral()
    }

    dependencies {
        classpath("io.gitlab.arturbosch.detekt:detekt-gradle-plugin:1.23.0")
    }
}

subprojects {
    apply plugin: 'io.gitlab.arturbosch.detekt'

    tasks.withType(Detekt).configureEach {
        reports {
            xml.required = false
            sarif.required = false
            html.required = true
        }
    }
}

```

<img width="299" alt="image" src="https://github.com/detekt/detekt/assets/106542177/b6a99bc8-2c0c-431e-ae6f-ceea5ea961c2">

## Your Environment

* Version of detekt used: Version 1.23.0 
* Version of Gradle used (if applicable): 8.0
* Gradle scan link (add `--scan` option when running the gradle task):
* Operating System and version: Apple M1 Macbook Pro Ventura 13.3.1
* Link to your project (if it's a public repository):  [DetektReportsReproducer.zip](https://github.com/detekt/detekt/files/11562916/DetektReportsReproducer.zip).

**Comment by 3flex** [comment]
Is this only happening when you add to classpath via build script? What if you use the `plugins {}` block?

**Comment by TWiStErRob** [comment]
TL;DR: WAI (this is how Gradle classloaders work with applied Groovy script files).

---

I added some logging:
app/build.gradle:
```gradle
println(project)
afterEvaluate { println("afterEvalute $it") }
```
detekt.gradle:
```gradle
subprojects {
    apply plugin: 'io.gitlab.arturbosch.detekt'
    println("subprojects: $it")
    afterEvaluate { println("subprojects.afterEvaluate $it") }
    tasks.withType(Detekt).configureEach {
        println("subprojects.configureEach $it")
```

output:
```console
> Configure project :
subprojects: project ':app'
subprojects: project ':mylibrary'

> Configure project :app
project ':app'
subprojects.afterEvaluate project ':app'
afterEvalute project ':app'

> Configure project :mylibrary
subprojects.afterEvaluate project ':mylibrary'
```

There are two things to note here:
1. there's no log line for `configureEach`! I think this will be the solution. Probably classpath conflict.
2. The order of execution is:
    ```
    subprojects: project ':app'
    project ':app'
    subprojects.afterEvaluate project ':app'
    afterEvalute project ':app'
    ```
    Which is a bit counter-intuitive, but it's just how Gradle works. The consequence here is that a plugin application in the more specific `build.gradle` may register an `afterEvaluate` block, which might overwrite your setup in the `subproject` block. This is more applicable to AGP setup, I don't think this is what's happening here.

Let's look at 1. a bit more. You applied
```
plugins { id "io.gitlab.arturbosch.detekt" version "1.23.0" apply false }
classpath("io.gitlab.arturbosch.detekt:detekt-gradle-plugin:1.23.0")
```
in two different files. If we remove either, some error happens.

Adding `println(io.gitlab.arturbosch.detekt.Detekt.class)`:
```
> Configure project :
subprojects: project ':app'
class io.gitlab.arturbosch.detekt.Detekt

> Configure project :app
project ':app'
class io.gitlab.arturbosch.detekt.Detekt
```
doesn't tell us much.

`Detekt.class.protectionDomain.codeSource.location` also gives the same `gradle/caches/jars-9/3842d3315a8876c1da5d384624c3ec8c/detekt-gradle-plugin-1.23.0.jar`

`Detekt.class.classLoader`, now we're getting somewhere:
```console
> Configure project :
subprojects: project ':app'
VisitableURLClassLoader(ClassLoaderScopeIdentifier.Id{coreAndPlugins:settings[:]:buildSrc[:]:script-file:/T:/DetektReportsReproducer/detekt.gradle(export)})

> Configure project :app
project ':app'
VisitableURLClassLoader(ClassLoaderScopeIdentifier.Id{coreAndPlugins:settings[:]:buildSrc[:]:root-project[:](export)})
```

Adding
```gradle
subprojects {
    apply plugin: 'io.gitlab.arturbosch.detekt'
    tasks.configureEach {
        def clz = it.class.superclass // Get the parent of the generated _Decorated classes which extend the real class at runtime.
        println("$it ${clz} (${clz == io.gitlab.arturbosch.detekt.Detekt.class}) (${clz.classLoader})")
    }
```
shows
```console
> Configure project : // detekt.gradle
VisitableURLClassLoader(ClassLoaderScopeIdentifier.Id{coreAndPlugins:settings[:]:buildSrc[:]:script-file:/T:/DetektReportsReproducer/detekt.gradle(export)})

> Configure project :app // app/build.gradle
VisitableURLClassLoader(ClassLoaderScopeIdentifier.Id{coreAndPlugins:settings[:]:buildSrc[:]:root-project[:](export)})
task ':app:detekt' class io.gitlab.arturbosch.detekt.Detekt (false) (VisitableURLClassLoader(ClassLoaderScopeIdentifier.Id{coreAndPlugins:settings[:]:buildSrc[:]:root-project[:](export)}))
```
which means that while both classes are the `Detekt` class, they're actually not equal. This means that `withType` doesn't have a chance to match the task. This is why we didn't see `configureEach`.

How to solve this quickly? 🤷‍♂️

When I first encountered this, I started using `buildSrc`, because it's not worth messing about with Gradle's multiple classloaders, move everything to one. Here's how:

1. Remove `plugins` block from `build.gradle`
2. Remove `buildscript` block from `detekt.gradle`
3. Create `buildSrc/build.gradle` with the following contents:
```gradle
repositories { // same as settings.gradle's pluginManagement.repositories
    mavenCentral()
    google()
}
dependencies { // from build.gradle (list all plugins used in the project)
    // plugins { id "<id>" version "<ver>" } -> implementation("<id>:<id>.gradle.plugin:<ver>")
    implementation("com.android.application:com.android.application.gradle.plugin:8.1.0-beta03")
    implementation("org.jetbrains.kotlin.android:org.jetbrains.kotlin.android.gradle.plugin:1.8.10")
    implementation("com.android.library:com.android.library.gradle.plugin:8.1.0-beta03")
    implementation("io.gitlab.arturbosch.detekt:io.gitlab.arturbosch.detekt.gradle.plugin:1.23.0")
}
```
and the output:
```console
> Configure project :
subprojects: project ':app'
VisitableURLClassLoader(ClassLoaderScopeIdentifier.Id{coreAndPlugins:settings[:]:buildSrc[:](export)})

> Configure project :app
project ':app'
VisitableURLClassLoader(ClassLoaderScopeIdentifier.Id{coreAndPlugins:settings[:]:buildSrc[:](export)})
task ':app:detekt' class io.gitlab.arturbosch.detekt.Detekt (true) (VisitableURLClassLoader(ClassLoaderScopeIdentifier.Id{coreAndPlugins:settings[:]:buildSrc[:](export)}))
subprojects.configureEach task ':app:detekt'
```
So the classLoaders match, the configureEach works, and the expected behavior is triggered.

Notice that while using `buildSrc`, it doesn't necessarily mean moving to precompiled plugins, it only exists to set up the classpath.

---

This still looks a bit hacky, nowadays precompiled script plugins are all the rage: have a look at https://github.com/TWiStErRob/net.twisterrob.colorfilters, the key places are `settings.gradle.kts`>`includeBuild` and `gradle/plugins/src/main/kotlin/`.

**Comment by ahibrahimleague** [comment]
Thanks for going through the rabbit hole on my behalf 🙏 

Now you've explained the problem, it makes sense why this didn't work, since the class loaders are different. I've seen a few issues  opened related to Detekt plugin being applied in buildscripts, I guess this information might be useful to be mentioned in the Detekt docs as a common pitfall that devs have to take care of.

**Comment by ahibrahimleague** [comment]
Closing ... as the most straightforward solution is basically to move configuring Detekt tasks inside the buildSrc directory.
